hail/src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -90,6 +90,23 @@ final case class Cast(v: IR, _typ: Type) extends IR
 final case class NA(_typ: Type) extends IR { assert(!_typ.required) }
 final case class IsNA(value: IR) extends IR
 
+object Coalesce {
+  def unify(values: Seq[IR], unifyType: Option[Type] = None): Coalesce = {
+    require(values.nonEmpty)
+    val t1 = values.head.typ
+    if (values.forall(_.typ == t1))
+      Coalesce(values)
+    else {
+      val t = unifyType.getOrElse(t1.deepOptional())
+      Coalesce(values.map(PruneDeadFields.upcast(_, t)))
+    }
+  }
+}
+
+final case class Coalesce(values: Seq[IR]) extends IR {
+  require(values.nonEmpty)
```

Comment:
I feel like this should just be in `Typecheck`

Reply:
We'll crash long before getting to typecheck if we try to construct a Coalesce without any values, and having the stack trace appropriately localized seems like a plus.

Reply:
hm, ok. It seems weird to me that there are a lot of things we do before typechecking. The IR shouldn't be valid if it ultimately won't typecheck and I feel like you'd need to know that first, aside from just having one place that describes the IR beyond the Scala type. Probably out of the scope of this PR to consider that though.

Reply:
we validate the IR a bunch of times, but an invalid Coalesce node in this way would crash on `InferType`, which is called before the node is checked itself in cases like:
`ArrayFilter(a, 'foo', Coalesce())`.

This node checks that its predicate has `typ == TBoolean` before recursively checking the children.

Reply:
Makes sense, thanks!